### ISSUE TITLE
Expand only on ImageButton click

### ISSUE THREAD

**Issue by rsd-raul** [mikepenz/FastAdapter#157]
Good evening Mike, I was wondering how to expand the SubItems only when clicking on the ImageButton.

My idea is to expand with the ImageButton onClickListener and use the Item onClickListener for another functionality, not sure if that´s possible, I´m assuming it is.

Btw: As always, brilliant work, I´m more than happy after making the switch to FastAdapter :D

**Comment by mikepenz** [comment]
@rsd-raul it is not possible with the included `ExpandableItem` but should be really easy to implement. 

In general the following changes have to be done:
remove the listener here (this is currently used to replace the default behavior): 
https://github.com/mikepenz/FastAdapter/blob/develop/app/src/main/java/com/mikepenz/fastadapter/app/items/ExpandableItem.java#L121

add the listener on the `ImageButton` with the functionality as the listener above had.

**Comment by mikepenz** [comment]
You need more help?

**Comment by rsd-raul** [comment]
Sorry for the late reply Mike, I didn't have time have any time left yesterday :S

I tried removing the listener as you sugested, but the main problem I have is implementing the expand behaviour.

I have this:
````
viewHolder.expandIcon.setOnClickListener(new View.OnClickListener() {
    @Override
    public void onClick(View view) {
        viewHolder.itemView.performClick();
    }
});
````

And this:
````
@Override
public FastAdapter.OnClickListener<TaskDetailItem> getOnItemClickListener() {
    return new FastAdapter.OnClickListener<TaskDetailItem>() {
        @Override
        public boolean onClick(View v, IAdapter<TaskDetailItem> adapter, TaskDetailItem item, int pos) {

            // If the item has sub-items rotate the icon 180º
            boolean moreInfo = false;
            if (item.getSubItems() != null) {
                int rotation = item.isExpanded() ? 180 : 0;

                ViewCompat.animate(v.findViewById(R.id.expandButton)).rotation(rotation).start();

                moreInfo = true;
            }

            // And if the item has a custom onClickListener call it
            return mOnClickListener != null ? mOnClickListener.onClick(v, adapter, item, pos) : moreInfo;
        }
    };
}
````

That makes the ````ImageButton```` capable of expanding, but clicking the item will also expand.

I´ve tried doing:
````
@Override
public FastAdapter.OnClickListener<TaskDetailItem> getOnItemClickListener() {
    return null;
}
````
But it still expands.

I'm going horribly wrong somewhere, right?


---

Using a boolean value I can manage to execute a statement only on the ````ImageButton````click, but I dont really know what to modify to block the expantion if the click is coming from the item:
````
viewHolder.expandIcon.setOnClickListener(new View.OnClickListener() {
    @Override
    public void onClick(View view) {
        mExpandButton = true;
        viewHolder.itemView.performClick();
    }
});
`````
````
boolean mExpandButton = true;`
````
`````
@Override
public FastAdapter.OnClickListener<TaskDetailItem> getOnItemClickListener() {
    return new FastAdapter.OnClickListener<TaskDetailItem>() {
        @Override
        public boolean onClick(View v, IAdapter<TaskDetailItem> adapter, TaskDetailItemitem, int pos) {

            // If the detail has more info rotate the icon 180º
            boolean moreInfo = false;
            if (item.getSubItems() != null) {
                int rotation = item.isExpanded() ? 180 : 0;

                ViewCompat.animate(v.findViewById(R.id.expandButton)).rotation(rotation).start();

                moreInfo = true;
            }

            if(expandButton) {
                Toast.makeText(context, "Testing", Toast.LENGTH_SHORT).show();
                expandButton = false;
            }

            // And if the item has a custom onClickListener call it
            return mOnClickListener != null ? mOnClickListener.onClick(v, adapter, item, pos) : moreInfo;
        }
    };
}
````

Thanks for the help :D

**Comment by rsd-raul** [comment]
I "fixed" it by surrounding everything but the ````ImageButton```` in another layout, it's not ideal but it works, plus I'm showing only a few elements :S

The layout will be something like:

````
Relative
    Linear                         <= customBehaviour();
        ImageButton
        TextView
    /Linear
    ImageButton                    <= viewHolder.itemView.performClick();
/Relative
````

**Comment by Rainer-Lang** [comment]
Could you even submit/pr your solution? Maybe as another example? Would be great!

**Comment by rsd-raul** [comment]
Hi Rainer, you mean my solution or Mike's?

**Comment by Rainer-Lang** [comment]
Yours is working, so yours. 
Did you try Mike's solution?

**Comment by mikepenz** [comment]
@rsd-raul it was not 100% possible with the previous release. I have now added a new feature to allow this. I also added this branch: https://github.com/mikepenz/FastAdapter/tree/feature/solution_issue_157 with the changes needed to get this proper working.

**Comment by rsd-raul** [comment]
Hi Mike, thanks for that, I followed the changes and adapted my code to it, but nothing happened, probably I have something different/missing and that´s why it´s misbehaving... Not really sure, maybe Rainer can try it and give some feedback.

On the other hand, maybe for this particular implementation is a little overkill, I mean, adding a new layout it´s not ideal but it´s not end of the world bad, plus, my list is 5/6 items tops, it shouldn't affect performance.

If I finally decide to make my main items expandable in the dashboard... Which is bound to happen... I'll make them using the new functionality, but it will help if it was simplified/foolproof (like everything else), maybe an optional:
````
.withExpansionItem(R.id.view)
````
And handling the click as you already handle the itemView clicks will be the ideal solution.

Ps: on the other hand, just noticed you make your fastAdapters with <IItem>, which means I can save a few lines in Dagger... Yey :dancer:

**Comment by rsd-raul** [comment]
![capture](https://cloud.githubusercontent.com/assets/6963554/16666884/83793084-4481-11e6-92d4-e8b9419b09d7.PNG)

Didn´t knew that was a thing xD

How cool is that?

**Comment by mikepenz** [comment]
@rsd-raul for the separate branch I have just modified the existing `ExpandableItem` you should copy and paste this item and do the modifications as in the commit + you should make it `autoExpanding` false:
https://github.com/mikepenz/FastAdapter/commit/74c70366c6b3b5b672bb3c042ef5b9b285554bd3#diff-e3cb79b0c8d0b1c3872031cefbdf5cf9R87

Use `IItem` as type for your adapter if you have multiple types of items, if you have just one you should use the specific type. 

Oh cool. You use private github repos?

**Comment by rsd-raul** [comment]
I tried that and also modified the fragment who handles the adapter to set the ````ClickListenerHelper```` as I saw you do in the commit.

I´ll probably download the branch when I get to it, I´m sure it works like a charm and I´m just missing something really small, but I have limited time this days so I´m trying to avoid perfection and focus on getting things done.

Ps: I finally removed the extra layout and set listeners to icons and text, behaves the same way, it´s more modular and I save a ````findById()```` and the layout by reusing the elements, so there is that :D

Regarding the repos, it depends, right now I´m using mostly SVN so I´m used to the "private" stuff, reason why for big projects and work-in-progress ones I tend to do the same.
